equational_theories/Closure.lean

```diff
@@ -0,0 +1,247 @@
+
+def Edge.lhs : Edge → String := Implication.lhs ∘ get
+
+def Edge.rhs : Edge → String := Implication.rhs ∘ get
+
+inductive Outcome
+  | explicit_proof_true
+  | implicit_proof_true
+  | explicit_conjecture_true
+  | implicit_conjecture_true
+  | unknown
+  | implicit_conjecture_false
+  | explicit_conjecture_false
+  | implicit_proof_false
+  | explicit_proof_false
+  deriving Repr, DecidableEq, Hashable, Lean.ToJson, Lean.FromJson
+
+instance : Inhabited Outcome where
+  default := .unknown
+
+instance : ToString Outcome where
+  toString := toString ∘ repr
+
+def Outcome.implicit_theorem : Bool → Outcome
+  | true => implicit_proof_true
+  | false => implicit_proof_false
+
+def Outcome.explicit_theorem : Bool → Outcome
+  | true => explicit_proof_true
+  | false => explicit_proof_false
+
+def Outcome.implicit_conjecture : Bool → Outcome
+  | true => implicit_conjecture_true
+  | false => implicit_conjecture_false
+
+def Outcome.explicit_conjecture : Bool → Outcome
+  | true => explicit_conjecture_true
+  | false => explicit_conjecture_false
+
+def Outcome.isExplicit : Outcome → Bool
+  | explicit_proof_true => true
+  | implicit_proof_true => false
+  | explicit_conjecture_true => true
+  | implicit_conjecture_true => false
+  | unknown => false
+  | implicit_conjecture_false => false
+  | explicit_conjecture_false => true
+  | implicit_proof_false => false
+  | explicit_proof_false => true
+
+def Outcome.isProven : Outcome → Bool
+  | explicit_proof_true => true
+  | implicit_proof_true => true
+  | explicit_conjecture_true => false
+  | implicit_conjecture_true => false
+  | unknown => false
+  | implicit_conjecture_false => false
+  | explicit_conjecture_false => false
+  | implicit_proof_false => true
+  | explicit_proof_false => true
+
+def Outcome.isTrue : Outcome → Bool
+  | explicit_proof_true => true
+  | implicit_proof_true => true
+  | explicit_conjecture_true => true
+  | implicit_conjecture_true => true
+  | unknown => false
+  | implicit_conjecture_false => false
+  | explicit_conjecture_false => false
+  | implicit_proof_false => false
+  | explicit_proof_false => false
+
+partial def dfs1 (graph : Array (Array Nat)) (vertex : Nat) (vis : Array Bool) (order : Array Nat) :
+    Array Bool × Array Nat := Id.run do
+  let mut vis1 := vis.set! vertex true
+  let mut ord := order
+  for v in graph[vertex]! do
+    unless vis1[v]! do
+      (vis1, ord) := dfs1 graph v vis1 ord
+  pure (vis1, ord.push vertex)
+
+partial def dfs2 (graph : Array (Array Nat)) (vertex : Nat) (component : Array Nat) (component_id : Nat) :
+    Array Nat := Id.run do
+  let mut comp := component.set! vertex component_id
+  for v in graph[vertex]! do
+    if component[v]! == 0 then
+      comp := dfs2 graph v comp component_id
+  pure comp
+
+def Bitset := Array UInt64
+
+def Bitset.toArray : Bitset → Array UInt64 := id
+
+instance : Inhabited Bitset := ⟨#[]⟩
+
+def Bitset.mk (n : Nat) : Bitset := Array.mkArray ((n + 63) >>> 6) 0
```

Comment:
What is going on here? Why shift by 6?

Reply:
This is a bitset (https://en.cppreference.com/w/cpp/utility/bitset), it represents a vector of booleans by packing them to `UInt64`, which makes computing a union 64 times faster, as we can use bitwise or.

Reply:
This would make a good comment in the code.

Reply:
Just to clarify this: if one is using the `conjecture` keyword, does one also need to add an explicit `@[equational_result]` tag, or is this automatic?

Reply:
Yes, I also updated `CONTRIBUTING.md` to mention that.